contracts/src/libraries/PlonkVerifier2.sol

```diff
@@ -0,0 +1,927 @@
+
+            // challenge: v
+            {
+                // preimage len: state(0x20) + transcript(0x140)
+                mstore(add(dataPtr, 0x140), keccak256(statePtr, 0x160))
+                // update new state (by updating state pointer)
+                statePtr := add(dataPtr, 0x140)
+                // empty transcript
+                dataPtr := add(statePtr, 0x20)
+                // (mod R_MOD) to get challenge
+                mstore(add(res, 0xc0), mod(mload(statePtr), R_MOD))
+            }
+
+            let zetaPtr := mload(add(proof, 0x160))
+            mstore(dataPtr, mload(zetaPtr)) // zeta.x
+            mstore(add(dataPtr, 0x20), mload(add(zetaPtr, 0x20))) // zeta.y
+            let zetaOmegaPtr := mload(add(proof, 0x180))
+            mstore(add(dataPtr, 0x40), mload(zetaOmegaPtr)) // zetaOmega.x
+            mstore(add(dataPtr, 0x60), mload(add(zetaOmegaPtr, 0x20))) // zetaOmega.y
+
+            // challenge: u
+            {
+                // preimage len: state(0x20) + transcript(0x80)
+                let hash := keccak256(statePtr, 0xa0)
+                // (mod R_MOD) to get challenge
+                mstore(add(res, 0xe0), mod(hash, R_MOD))
+            }
+        }
+    }
+
+    function verify(
+        IPlonkVerifier.VerifyingKey memory vk,
+        uint256[8] memory publicInput,
+        IPlonkVerifier.PlonkProof memory proof
+    ) external view returns (bool success) {
+        _validateProof(proof);
+
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[0]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[1]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[2]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[3]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[4]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[5]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[6]));
+        BN254.validateScalarField(BN254.ScalarField.wrap(publicInput[7]));
+
+        Challenges memory chal = _computeChallenges(vk, publicInput, proof);
+        Poly.EvalDomain memory domain = Poly.newEvalDomain(vk.domainSize);
+        Poly.EvalData memory evalData = Poly.evalDataGen(domain, chal.zeta, publicInput);
+
+        uint256 linPolyConstant = _computeLinPolyConstantTerm(chal, proof, evalData);
+
+        assembly {
+            // credit: error function from Succinct's SP1 code
+            function error_verify() {
+                let ptError := mload(0x40)
+                mstore(ptError, ERROR_STRING_ID) // selector for function Error(string)
+                mstore(add(ptError, 0x4), 0x20)
+                mstore(add(ptError, 0x24), 0xc)
+                mstore(add(ptError, 0x44), "error verify")
+                revert(ptError, 0x64)
+            }
+            function error_pairing() {
```

Comment:
This function is used only once, why not inline it?

Reply:
we could, but i think this is clearer and easier to audit, and won't cost much extra, just 2~3 more opcode, (push this function selector to stack, jump, then pop off (~10 gas)